### PR TITLE
Adding data and UCDs

### PR DESCRIPTION
**Pull request by AlecThomson**
Hey @Cameron-Van-Eck,

I've done a bit of a re-work here to store the 'standard' part of the RMTable standard in separate JSON files outside of the code itself. These files should be packaged correctly, hence some of the file moving, and end up where the code can find it. I've also added checks on version strings for if/when things get updated.

I've started to add UCD entries for some of the columns. It's going to be a bit of a slog, and thanks to their compound nature we may want to combo some together for our columns. There are also some columns that don't really seem to fit any UCDs. I'm also still not sure exactly how the metadata part of UCDs work, either. For reference, I've included the list of defined UCDs in [ucd1p-words.txt](https://github.com/CIRADA-Tools/RMTable/compare/master...AlecThomson:ucd?expand=1#diff-f1b710933827cc67ef21084ff0a1f9cabf820aa95b4e4f092b5c4ba39ca71e15).

Let me know what you think of this.

Cheers,
Alec

### CODE REVIEW

**Comment by AlecThomson** [comment]
P.S. I still haven't added in any actual use of the UCD data yet. I know its possible for VOtables (and their astropy handlers) but I'm not sure about Tables in general nor FITS versions. I'll do some more digging.

**Comment by Cameron-Van-Eck** [comment]
Hi @AlecThomson:

Thanks a lot for working on this. I really like the JSON implementation of the standard -- that should make it a lot easier to make auto-documentation for any future changes to the standard (e.g., adding new standard strings), which is on my to-do list for this week.

Do you think it would be possible to move the reading of the JSON from the class initialization to the importing of the module? I know it's probably a trivial performance hit, but it mildly annoys me to be reading from disk every time we initialize an RMTable.

I also did a little work into adding UCDs to the RMTable standard. I committed a version with partial UCD support, then withdrew it (commit [2072e01](https://github.com/CIRADA-Tools/RMTable/commit/2072e012804d226de2eb56e4a52cd2b1d329afa0)): https://github.com/CIRADA-Tools/RMTable/blob/f8354fc5176280a1da3b42342d4961b548da6245/rmtable.py
You can see the UCDs I chose there if you want to compare them with yours. From a quick inspection, the main difference I see is our treatment of error columns (stat.error is a primary keyword, so it has to come first). You can also see my incomplete integration of the UCDs into the rest of the code.

The reason I withdrew it is because I got frustrated with the problem of trying to maintain/progate UCDs through read and write operations, particularly with FITS format. The Astropy FITS read/write tools can handle units somewhat gracefully, but have no built-in methods of handling UCDs. I didn't even try with the TSV format. I also ran into complications with some operations (removing columns, etc) that were probably solvable but tedious. I decided it wasn't worth the effort in the short term, so I went for "minimum possible compliance" to get VOTable functionality to work. If you want to continue on it, be my guest (I'm drowning in other work at the moment). I think basic read/write for VOTables is sufficient to address the referee's comments, with UCD support being a nice stretch goal that I currently can't work on.

Possibly relevant: a slightly alternative approach that I experimented with (but similarly withdrew) for PolSpectra was storing the UCDs as a dictionary rather than a list (this gives advantages in terms of keeping the association between columns name and UCD closer/safer; I'm considering doing the same with the limits and units lists): https://github.com/CIRADA-Tools/PolSpectra/commit/ccd39f689763491bf01fd33e22a62d29faf159d7

Cheers,
Cameron

**Comment by AlecThomson** [comment]
Hey @Cameron-Van-Eck,

Thanks for pointing me to those commits. I've added in all those UCDs as you had them. I've also moved the reading of the JSON to the global space using `__xx__` to keep the namespace clean.

After some digging in the tools for converting between astropy votables and regular tables, I found that the ucd info is stored in the `meta` property of each column. By setting it there, writing to FITS and VOtable automagically works! Using the example as a test, everything seems to be working pretty well now. Hopefully this can be useful for polspectra as well.

Just out of interest, would it worth also exploring just using inheritance from astropy tables to give a more general ability to read/write? Although, that might look like a major overhaul that perhaps neither of us have the time to commit to right now!

**Comment by Cameron-Van-Eck** [comment]
This is looking good to me, I think. There's some fussy details to be added to some of the functions (e.g., add_column() should allow users to supply a UCD, append_to_table() needs to handle mergers gracefully, etc), but I can probably handle most of that myself if you don't want to. Let me know when/if you think you're done with this and I'll merge it in and deal with the remaining details.

Cheers,
Cameron

**Comment by Cameron-Van-Eck** [comment]
Also, forgot to say earlier, but I agree with you on how using inheritance would have been a good thing (and probably would have simplified thing enormously). Alas, at the time I started on this I didn't understand enough about classes (and probably still don't) to do that properly. The current framework is a bit of a kludge, but it seems like it should work well enough for the time being that I can't justify spending the time to do it better....

**Comment by AlecThomson** [comment]
Hey @Cameron-Van-Eck,

After sleeping on it, I was able to get inheritance to work! This makes reading/writing much easier, and also has the benefit of bringing along all the other benefits of astropy tables. The biggest change in behaviour is that the standard columns are now always created when indexing more than one column. I've updated the example to show how this works. But, it's trivially easy to convert to/from astropy tables and rmtables now, so this can be avoided. To my mind this behaviour, whilst potentially annoying at first, makes sense for the purpose of an RMTable. It's an extended astropy Table that enforces a _minimum_ set of columns.

e.g.
```python
from rmtable import RMTable
from astropy.table import Table
tab = Table.read('some_file')
rmtab = RMTable(tab)
# or
rmtab = RMTable.read('some_file')
tab = Table(rmtab) # or
tab = rmtab.to_table()
```

If you're not a fan of this change, feel free to ignore commits after [21a2ad0](https://github.com/CIRADA-Tools/RMTable/pull/4/commits/21a2ad0062ebe31f5caa1cc525d0d6dab4c39e00)

**Comment by Cameron-Van-Eck** [comment]
Hi @AlecThomson:

I really like the inheritance version. I feel like this is simplifying this a fair bit, and doing things in a way that is more proper.

But we've lost some functionality that I think is really important. I ended up making a Jupyter notebook to test things, so it's probably easier to post that than to describe everything. Link: https://www.dropbox.com/s/3gkk97x5vma7sgr/Behaviour%20testing.ipynb?dl=0

Long story short, there's some convenience functions that have been lost, and a few instances of weird behaviour. Are you willing to take a look and seeing which of these can be easily fixed?

Thanks,
Cameron

**Comment by AlecThomson** [comment]
Thanks for those check @Cameron-Van-Eck! I think I've gotten us there. Going through your list:
- round-trip read/write FITS without changes
- same, for VOTable format
This is now achieved. I've added the helper method `write_votable` to make things more friendly. Otherwise `write` can handle everything `astropy.table.Table` can. I tracked down the masking and apparent `--` values to `numpy.ma.MaskedArray`. This value is actually just used for display, but the underlying data are left intact.
```python
# The '--' strings are just what's displayed by np.ma.MaskedArray
# The data are still there!
# The masking just hides them from the user on print-like outputs
print(cat['pos_err'])
print(cat['pos_err'].data)
print(cat['pos_err'].data.data)
```
```
pos_err
  deg  
-------
     --
     --
     --
     --
     --
     --
     --
     --
     --
     --
    ...
     --
     --
     --
     --
     --
     --
     --
     --
     --
     --
Length = 55819 rows
[-- -- -- ... -- -- --]
[nan nan nan ... nan nan nan]
```
- same for TSV format (this isn't achieved in the main branch currently, because units and UCDs are lost...)
I've added a `read/write_tsv` helper. But UCDs and units are lost. We'd need to implement some kind of custom metadata solution for this.
- add columns simply, with units and UCDs
- remove columns simply
- correctly track the current unit/UCD contents of the table
- correctly track columns/units/ucds present (including additions and subtractions)
I've added `add/remove_column/s` to handle this. It mostly just uses the astropy method, but handles the `ucds` and `units` properties.
- extract subsets of tables. Row-wise subsets (i.e., keeping all columns) should preserve RMTable status, but probably not column-wise subsets (I think my own code doesn't do this entirely properly? The current main branch returns an RMTable object with fewer columns, which isn't really proper; at that point it should just be a Table probably.)
- verification should be something manually triggered by the user, not an automatic thing
I've removed the auto-checking in the `__init__` function, leaving it to the user to run them. This simplified things a lot, as it turns out `__init__` is called all over the place. Subsetting now returns a smaller `RMTable`. Separating row/column-wise splitting was looking to be nightmare. It doesn't seem too bad to have a cut down `RMTable` object, as it still is mostly an astropy Table with extra methods.

- need to be able to merge RMTables together (including dealing with differences in what columns are present (inner/outer/exact joins)).
After much pain, this also works.

This is after some rough testing by me, I have probably missed some things. We may want to convert your testing script into a proper CI testing script at some point.

**Comment by AlecThomson** [comment]
P.S. I just came across astropy's recommendation for their [ecsv format](https://docs.astropy.org/en/stable/io/ascii/ecsv.html#ecsv-format). This is supported out of the box by astropy Tables, and stores all the metadata in a header schema. With RMtables we could just do:
```python
cat.write('./test_table.ecsv')
readecsv=RMTable.read('test_table.ecsv')
```

**Comment by AlecThomson** [comment]
P.P.S, I see now in the ascii format nan values are written as `""` from the masked arrays but resolve to `nan` from the `astropy.io.fits` reading. Grr. We could maybe force masked arrays to become unmasked on write.

**Comment by AlecThomson** [comment]
Fixed the above with the `_unmask` method

**Comment by Cameron-Van-Eck** [comment]
Sorry again for the delay in looking at this. The changes are great; you've addressed most of the problems I brought up. There's just a few funny behaviours I want to point out for potential fixes:
- There's two add_columns() functions now. Presumably you meant to remove the 2nd one?
- When new (custom/non-standard) columns are added, the self.ucd and self.unit attributes don't seem to get updated.
- It's not necessarily intuitive to a user how to add a column with a unit and ucd. It would be nice to have those as explicit (but still optional) inputs to the add_column() function. The method I've tried so far, which is a bit cumbersome (particularly the UCD), looks like `newcol=at.Column(data=np.random.randint(1000,size=len(SGPS)),name='random2',unit='Jy',
                 meta={'ucd':'stat.error;meta.number'}); SGPS.add_column(newcol)`. Doing it this way causes ths units/UCDs to go into the table properly, but not populate the attributes as expected.
- added/custom columns without UCDs have a None for their UCD after round-trip (save and read). Not necessarily a problem, but it has the potential to break things that expect a string (would be better as empty string, but not necessarily important enough to spend time on).
- some of the new functions like write_votable() could use slightly longer docstrings to explain the units.
-the read_tsv() function seems a bit temperamental: it won't read the table I made during testing, and it won't read the existing .tsv consolidated catalog (with different errors on each). Are you able to round-trip tables with tsv format?

I like the ECSV idea. We basically already support it, by using the Table read/write functions, right? I think I won't change the paper to go into it, but will add it to the online documentation/wiki.

I also really like having the descriptions included in the json and the tables.

Sorry if this is a bit fussy. I think this is great. Fix whatever you like from that list, and I'll go ahead and merge it in (and start updating the documentation to match the new functions/behaviour). Next week is a mess for me (long weekend + department retreat), but hopefully I'll be able to get this done late next week or early week after (and then hopefully finish the last of the referee comments!)

Thanks,
Cameron

**Comment by AlecThomson** [comment]
Hey @Cameron-Van-Eck,

Ok brilliant! I've fixed the obviously bad bits of `add_column`, now there's just one of them! I've also added the optional kwargs to allow a user to add a ucd/unit on adding a column. Doing the same for `add_columns` would be trickier, so I've left that as-is for now.

If you're happy, I might leave the rest to you on your merge - which I hope isn't too painful...

**Comment by AlecThomson** [comment]
Caught some bugs after testing in the wild
